core/src/queue.rs

```diff
@@ -147,49 +147,84 @@ impl Queue {
         tx: VersionedAcceptedTransaction,
         wsv: &WorldStateView,
     ) -> Result<(), (VersionedAcceptedTransaction, Error)> {
-        if tx.is_in_future(self.future_threshold) {
-            Err((tx, Error::InFuture))
-        } else if let Err(e) = self.check_tx(&tx, wsv) {
-            Err((tx, e))
-        } else if self.txs.len() >= self.max_txs {
-            Err((tx, Error::Full))
-        } else {
-            let hash = tx.hash();
-            let entry = match self.txs.entry(hash) {
-                Entry::Occupied(mut old_tx) => {
-                    // MST case
-                    old_tx
-                        .get_mut()
-                        .as_mut_v1()
-                        .signatures
-                        .extend(tx.as_v1().signatures.clone());
-                    return Ok(());
+        match self.check_tx(&tx, wsv) {
+            Err(e) => Err((tx, e)),
+            Ok(MustUse(signature_check)) => {
+                // Get `txs_len` before entry to avoid deadlock
```

Comment:
Or, get the lock explicitly and drop it when done. Picking up and almost immediately dropping a mutex guard is not the worst but it's inefficient.

Reply:
I think dropping lock is not option here because we need to insert value into the `entry` just after check for size.